zcash_client_backend/src/scanning.rs

```diff
@@ -345,41 +340,106 @@ pub(crate) fn scan_block_with_runner<
-    // received notes. Thus, we allow it to be optional here, and only produce an error if we try
-    // to use it. `block.sapling_commitment_tree_size` is expected to be correct as of the end of
-    // the block, and we can't have a note of ours in a block with no outputs so treating the zero
-    // default value from the protobuf as `None` is always correct.
-    let mut sapling_commitment_tree_size = block
-        .chain_metadata
-        .as_ref()
-        .and_then(|m| {
-            if m.sapling_commitment_tree_size == 0 {
-                None
-            } else {
-                let block_note_count: u32 = block
-                    .vtx
-                    .iter()
-                    .map(|tx| {
-                        u32::try_from(tx.outputs.len()).expect("output count cannot exceed a u32")
-                    })
-                    .sum();
-                Some(m.sapling_commitment_tree_size - block_note_count)
-            }
-        })
-        .or_else(|| prior_block_metadata.map(|m| m.sapling_tree_size()))
-        .ok_or(ScanError::TreeSizeUnknown {
-            protocol: ShieldedProtocol::Sapling,
-            at_height: cur_height,
-        })?;
+    let initial_sapling_tree_size = prior_block_metadata.and_then(|m| m.sapling_tree_size());
+    let mut sapling_commitment_tree_size = initial_sapling_tree_size.map_or_else(
+        || {
+            block.chain_metadata.as_ref().map_or_else(
+                || {
+                    // If we're below Sapling activation, or Sapling activation is not set, the tree size is zero
+                    params
+                        .activation_height(NetworkUpgrade::Sapling)
+                        .map_or_else(
+                            || Ok(0),
+                            |sapling_activation| {
+                                if cur_height < sapling_activation {
+                                    Ok(0)
+                                } else {
+                                    Err(ScanError::TreeSizeUnknown {
+                                        protocol: ShieldedProtocol::Sapling,
+                                        at_height: cur_height,
+                                    })
+                                }
+                            },
+                        )
+                },
+                |m| {
+                    let sapling_output_count: u32 = block
+                        .vtx
+                        .iter()
+                        .map(|tx| tx.outputs.len())
+                        .sum::<usize>()
+                        .try_into()
+                        .expect("Sapling output count cannot exceed a u32");
+
+                    // The default for m.sapling_commitment_tree_size is zero, so we need to check
+                    // that the subtraction will not underflow; if it would do so, we given invalid
+                    // chain metadata for a block with Sapling outputs.
```

Comment:
```suggestion
                    // that the subtraction will not underflow; if it would do so, we were given
                    // invalid chain metadata for a block with Sapling outputs.
```